pkgs/_test_macros/lib/json_codable.dart

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) 2024, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:dart_model/dart_model.dart';
+import 'package:macro/macro.dart';
+import 'package:macro_service/macro_service.dart';
+
+/// A macro which adds a `fromJson(Map<String, Object?> json)` JSON decoding
+/// constructor to a class.
+class JsonCodable {
+  const JsonCodable();
+}
+
+final _jsonMapType =
+    '{{dart:core#Map}}<{{dart:core#String}}, {{dart:core#Object}}?>';
+
+class JsonCodableImplementation implements Macro {
+  @override
+  MacroDescription get description => MacroDescription(
+      annotation: QualifiedName(
+          uri: 'package:_test_macros/json_codable.dart', name: 'JsonCodable'),
+      runsInPhases: [2, 3]);
+
+  @override
+  Future<AugmentResponse> augment(Host host, AugmentRequest request) async {
+    return switch (request.phase) {
+      1 => phase1(host, request),
+      2 => phase2(host, request),
+      _ => AugmentResponse(augmentations: []),
+    };
+  }
+
+  Future<AugmentResponse> phase1(Host host, AugmentRequest request) async {
+    final target = request.target;
+    return AugmentResponse(augmentations: [
+      Augmentation(code: '''
+external ${target.code}.fromJson($_jsonMapType json);
+external $_jsonMapType toJson();
+   '''),
+    ]);
+  }
+
+  Future<AugmentResponse> phase2(Host host, AugmentRequest request) async {
+    final result = <Augmentation>[];
+
+    final target = request.target;
+    final model = await host.query(Query(target: target));
+    final clazz = model.uris[target.uri]!.scopes[target.name]!;
+    final initializers = <String>[];
+    for (final field
+        in clazz.members.entries.where((m) => m.value.properties.isField)) {
```

Comment:
Today, is it really easy to emit a reference to a member in code, but with this model the Member doesn't have the info required to do that at all - you have to cobble together a bunch of info about how you reached the member.

Can we reasonably add a `QualifiedName` to each member, and make these just lists instead of maps? And then you can easily emit references to the qualified names in code, given only the `Member`, and they will be guaranteed safe.